sympy/solvers/solveset.py

```diff
@@ -1121,3 +1124,597 @@ def linsolve(system, *symbols):
+                         'system is to be found.')
+
+    if hasattr(symbols[0], '__iter__'):
+        symbols = symbols[0]
+
+    try:
+        sym = symbols[0].is_Symbol
+    except AttributeError:
+        sym = False
+
+    if not sym:
+        raise ValueError('Symbols or iterable of symbols must be given as '
+                         'second argument, not type \
+            %s: %s' % (type(symbols[0]), symbols[0]))
+    polys = []
+    polys_expr = []
+    nonpolys = []
+    denominators = set()
+    for eq in system:
+        # Store denom expression if it contains symbol
+        denominators.update(_simple_dens(eq, symbols))
+        # try to remove sqrt and rational power
+        without_radicals = unrad(simplify(eq))
+        if without_radicals:
+            eq_unrad, cov = without_radicals
+            if not cov:
+                eq = eq_unrad
+        eq = eq.as_numer_denom()[0]
+        # this will remove sqrt if symbols are under it
+        poly = eq.as_poly(*symbols, extension=True)
+        if poly is not None:
+            polys.append(poly)
+            polys_expr.append(poly.as_expr())
+        else:
+            nonpolys.append(eq)
+
+    if len(symbols) == len(polys):
+        # If all the equations in the system is poly
+        if is_zero_dimensional(system):
+            try:
+                # try to solve 0 dimensional system using `solve_poly_system`
+                result = solve_poly_system(polys, *symbols)
+                # May be some extra soln is added because
+                # we did `poly = g.as_poly(*symbols, extension=True)`
+                # need to check that and remove if not a soln
+                result_update = S.EmptySet
+                for res in result:
+                    dict_sym_value = dict(list(zip(symbols, res)))
+                    if all(checksol(eq, dict_sym_value) for eq in system):
+                        result_update += FiniteSet(res)
+                return result_update
+            except NotImplementedError:
+                # Right now it doesn't fail for any polynomial system of
+                # equation if `solve_poly_system` fails then substitution
+                # method will handle it. pass it
+                pass
+
+        # positive dimensional system
+        # substitution method where new system is groebner basis of the system
+        basis = groebner(polys, symbols, polys=True)
+        new_system = []
+        for poly_eq in basis:
+            new_system.append(poly_eq.as_expr())
+        result = [{}]
+        result = substitution(
+            new_system, symbols, result, [],
+            denominators)
+
+    else:
+        # If alll the equations are not polynomial.
+        # Use substitution method for nonpoly + polys eq
+        result = substitution(
+            polys_expr + nonpolys, symbols, exclude=denominators)
+
+    if result:
+        return result
+    else:
+        return S.EmptySet
```

Comment:
Right now there is two cases [test1](https://github.com/sympy/sympy/pull/11111/files#diff-85baa04bbf4e1dfd9128782738e45424R1041), [test2](https://github.com/sympy/sympy/pull/11111/files#diff-85baa04bbf4e1dfd9128782738e45424R1224) I know when it returns EmptySet and it is correct also.

It doesn't check whether system is inconsistent or not . But `solveset_real`  ( or `solveset_complex` used in `substitution`) will return solution for at least some variables if there is soln. If `solveset_real`  (or or `solveset_complex` used in `substitution`) return `S.EmptySet` for all the variables present in symbols then it will return `S.EmptySet`. If `solveset` returns `ConditionSet` for all the variables then at that time `S.EmptySet` looks incorrect. What should it return in that case? I think `solveset` will solve for at least some variable and will not return `conditionset` all the time in `substitution` function.